### PR TITLE
LESS Conversion & Platform Upgrade

### PR DESCRIPTION
**Pull request by mindreeper2420**
This PR converts the platform to use LESS instead of SASS for CSS processing in order to both consume PatternFly (core) with extendibility in mind, as well as consume (and contribute) to various PatternFly and fabric8-ui widget repos.

As part of the PR, the following changes were made:
1. Convert SCSS files to LESS and update `*.component.ts` files to consume the new format
2. Convert imports, includes, functions and mixins to match LESS language features
3. Convert Webpack to build LESS files and consume the proper node modules
4. Upgrade PatternFly in `package.json` to  version 3.25.0, with all of it's required Bootstrap and Font Awesome dependencies
5. Restructure `main.less` to be the primary building block of the platform (see updates for LESS language features)
6. Begin to convert (em) and (rem) to (px), following work currently being done on Planner
7. Begin to revert platform back to a grid-based system, in order to properly build off of PatternFly and to match UX Design criteria

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: af1e41be7d2dca215855f29ce7c11747dfa00421
  * Your subject line is longer than 50 characters

 * Commit: f0be8119f99359775a4bdf1e3c11791fdd7d6389
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: dd4d81d046bcd003805352278fe4de0ce7488336
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

 * Commit: c9ed27442a439215a6f1ed93f74caf61b9f1b485
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: af1e41be7d2dca215855f29ce7c11747dfa00421
  * Your subject line is longer than 50 characters

 * Commit: f0be8119f99359775a4bdf1e3c11791fdd7d6389
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: dd4d81d046bcd003805352278fe4de0ce7488336
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

 * Commit: c9ed27442a439215a6f1ed93f74caf61b9f1b485
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: af1e41be7d2dca215855f29ce7c11747dfa00421
  * Your subject line is longer than 50 characters

 * Commit: f0be8119f99359775a4bdf1e3c11791fdd7d6389
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: dd4d81d046bcd003805352278fe4de0ce7488336
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**
  * Your subject line is longer than 50 characters

 * Commit: c9ed27442a439215a6f1ed93f74caf61b9f1b485
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b24ad5f88b3c280a3560c4245032f65384b1e5be
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

File: gulpfile.js [hunk 1]
```diff
@@ -15,13 +16,21 @@ const gulp = require('gulp'),
...
 function copyToDest(srcArr) {
   return gulp.src(srcArr)
     .pipe(gulp.dest(function (file) {
       return libraryDist + file.base.slice(__dirname.length); // save directly to dist
     }));
 }
 
+//Less compilation and minifiction - adopted from sass compilation, needs work
+gulp.task('transpile-less', function () {
+  return transpileLESS(appSrc + '/app/**/*.less');
```
Comment (dlabrecq) [inline gulpfile.js]: I suspect we want to revert theses changes as webpack is now performing the less conversion.

File: tsconfig.json [hunk 2]
```diff
@@ -11,8 +11,6 @@
     "strictNullChecks": false,
     "noUnusedLocals": false,
     "baseUrl": "./src",
-    "paths": {
-    },
```
Comment (dlabrecq) [inline tsconfig.json]: We can revert these changes.

File: src/assets/stylesheets/shared/_layout.scss [hunk 3]
```diff
@@ -1,371 +0,0 @@
...
-      padding-left: #{$i}px !important;
-    }
-    .padding-bottom-#{$i} {
-      padding-bottom: #{$i}px !important;
-    }
-    .padding-right-#{$i} {
-      padding-right: #{$i}px !important;
-    }
-  }
-}
```
Comment (dlabrecq) [inline src/assets/stylesheets/shared/_layout.scss]: Do we want to remove this or perhaps it has moved?

FYI, I'm using the following Less conversion in the patternfly-ng repo.

.padding-loop (@i) when (@i > 0) {
  .padding-@{i} {
    padding: ~"@{i}px" !important;
  }
  .padding-top-@{i} {
    padding-top: ~"@{i}px" !important;
  }
  .padding-left-@{i} {
    padding-left: ~"@{i}px" !important;
  }
  .padding-bottom-@{i} {
    padding-bottom: ~"@{i}px" !important;
  }
  .padding-right-@{i} {
    padding-right: ~"@{i}px" !important;
  }
  .padding-loop(@i - 1);
}
.padding-loop(15);

File: src/assets/stylesheets/shared/_layout.scss [hunk 4]
```diff
@@ -1,371 +0,0 @@
...
-    .padding-left-standard-offset {
-      padding-left: $grid-gutter-width/2 !important;
-    }
-    .padding-right-standard-offset {
-      padding-right: $grid-gutter-width/2 !important;
-    }
-    .padding-bottom-standard-offset {
-      padding-bottom: $grid-gutter-width/2 !important;
-    }
-}
```
Comment (dlabrecq) [inline src/assets/stylesheets/shared/_layout.scss]: Do we want to remove this?

FYI, I'm using the following Less conversion in the patternfly-ng repo.

.margin-loop (@i) when (@i > 0) {
  .margin-v {
    margin: ~"@{i}px" !important;
  }
  .margin-top-@{i} {
    margin-top: ~"@{i}px" !important;
  }
  .margin-left-@{i} {
    margin-left: ~"@{i}px" !important;
  }
  .margin-bottom-@{i} {
    margin-bottom: ~"@{i}px" !important;
  }
  .margin-right-@{i} {
    margin-right: ~"@{i}px" !important;
  }
  .margin-loop(@i - 1);
}
.margin-loop(15);

File: src/app/team-membership-dialog/team-membership-dialog.component.less [hunk 5]
```diff
@@ -26,28 +26,26 @@
...
-      padding: 0;
+    >div { padding: 0;
       label {
-        width: $width100 - 10%;
+        width: calc(100% - 10%);
         padding-left: em(20);
         font-weight: 500;
         span {
-          width: $width100;
+          .width-100;
```
Comment (dlabrecq) [inline src/app/team-membership-dialog/team-membership-dialog.component.less]: I believe this should just be 100%. The width-100 class is really only used a shortcut used in the HTML pages.

File: src/app/team-membership-dialog/team-membership-dialog.component.less [hunk 6]
```diff
@@ -26,28 +26,26 @@
...
     list-style: none;
     padding: 0;
-    border: 1px solid $color-pf-black-400;
-    background: $color-pf-white;
+    border: 1px solid;
+    border-color: @color-pf-black-400;
+    background: @color-pf-white;
     li {
-      width: $width100;
+      .width-100;
```
Comment (dlabrecq) [inline src/app/team-membership-dialog/team-membership-dialog.component.less]: I believe this should just be 100%. The width-100 class is really only used a shortcut used in the HTML pages.

File: src/app/team-membership-dialog/team-membership-dialog.component.less [hunk 7]
```diff
@@ -1,11 +1,11 @@
-@import '../../assets/stylesheets/base';
+@import (reference) '../../assets/stylesheets/_base';
 .createTeamMembershipBlock {
   font-size: rem(10);
   padding: 0;
   >div {
     margin-top: em(15);
     label {
-      width: $width100;
+      .width-100;
```
Comment (dlabrecq) [inline src/app/team-membership-dialog/team-membership-dialog.component.less]: I believe this should just be 100%. The width-100 class is really only used a shortcut used in the HTML pages. 

Due to the width-100 name, it's not likely to be a variable that will ever change from 100%.

File: tsconfig.json (hunk 2 above)
Comment (mindreeper2420) [inline tsconfig.json]: so don't delete that?

Comment (dlabrecq) [inline tsconfig.json]: Didn't notice there wasn't anything in the body. Please disregard my comment.

File: src/assets/stylesheets/shared/_layout.scss (hunk 3 above)
Comment (mindreeper2420) [inline src/assets/stylesheets/shared/_layout.scss]: @dlabrecq this was commented out, but I have added your conversion in its place.

File: src/assets/stylesheets/shared/_layout.scss (hunk 4 above)
Comment (mindreeper2420) [inline src/assets/stylesheets/shared/_layout.scss]: @dlabrecq this was commented out, but I have added your conversion in its place.

File: src/app/team-membership-dialog/team-membership-dialog.component.less (hunk 7 above)
Comment (mindreeper2420) [inline src/app/team-membership-dialog/team-membership-dialog.component.less]: agree - updated throughout the components

File: src/app/team-membership-dialog/team-membership-dialog.component.less (hunk 6 above)
Comment (mindreeper2420) [inline src/app/team-membership-dialog/team-membership-dialog.component.less]: agree - updated throughout the components

File: src/app/team-membership-dialog/team-membership-dialog.component.less (hunk 5 above)
Comment (mindreeper2420) [inline src/app/team-membership-dialog/team-membership-dialog.component.less]: agree - updated throughout the components

File: gulpfile.js (hunk 1 above)
Comment (mindreeper2420) [inline gulpfile.js]: changes reverted in next commit

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: b24ad5f88b3c280a3560c4245032f65384b1e5be
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 93b92935654bfb1e7ad62640b8a285b47c16332a
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by dlabrecq** [comment]
The latest changes are looking good. Just a couple nits -- probably ok to address these as separate issues?

1. fabric8-ui-mixins.less and fabric8-ui-variables.less are empty. It's not clear when to use these files instead of _custom-mixins.less and _custom-variables.less?

2. The develpoer poll widget doesn't have the right spacing between buttons.

screen shot 2017-06-01 at 11 59 51 am

3. The layout for the cancel/update buttons on the edit profile page is broken.

screen shot 2017-06-01 at 12 00 55 pm

4. The stack report widget (on the dashboard) no longer has a default selection. This also happens when creating a space -- the default template appears to be selected, but not shown in the drop down initially.

screen shot 2017-06-01 at 1 11 37 pm

5. The planner backlog has an extra large margin between the tree list and the side nav -- it doesn't appear on the board view.

screen shot 2017-06-01 at 1 13 47 pm

6. The icon spacing on the codebases page is off

screen shot 2017-06-01 at 1 16 02 pm

7. The space wizard layout is a little off. Can't click on the first two getting started options.

screen shot 2017-06-01 at 1 20 00 pm

**Comment by dlabrecq** [comment]
8. The space wizard no longer highlights getting started options.

9. The space wizard steps are broken.

screen shot 2017-06-01 at 1 28 34 pm

**Comment by dlabrecq** [comment]
10. Planner backlog and borad do not fill the entire page height.

screen shot 2017-06-01 at 1 41 19 pm

**Comment by mindreeper2420** [comment]
@dlabrecq Comments (with some fixes) below:

1 - removed the empty files, as I do not see them referenced anywhere
2 - I re-routed the manner in which `_utilities.less` is consumed, and broke the formulas out into their own file. Now the padding and margin loops work as expected.
3 - fixed, but not as part of (2). I changed the HTML/CSS to use the `:last-child` selector, extending off a base style.
4 - you crossed out :-)
5 - I believe this has to do with the `@for` expression that was used in the SCSS file. LESS does not support these type of expressions (if they do, I can not see an equation that I can use), so there are issues with the flex overlay. If we remove flex, then this won't matter :-) Planner is already working on removing flex.
6 - fixed as part of (2)
7 - the wizard is all sorts of messed up after the 2nd screen. I suggest a re-write of the HTML/CSS around it.
8 - see (7)
9 - see (7)
10 - I can fix this with some hacks, but Planner is removing flex and reverting to the grid based system. Platform will need to do the same.

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 93b92935654bfb1e7ad62640b8a285b47c16332a
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 12276778ca95579d065adf09b53c34a60e3598bf
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 12276778ca95579d065adf09b53c34a60e3598bf
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: 12276778ca95579d065adf09b53c34a60e3598bf
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by GitCop** [comment]
There were the following issues with your Pull Request

 * Commit: eb2d35792fe2d1ccadffb19ddf635ae7fe426c80
  * Your subject line is longer than 50 characters

Guidelines are available at https://github.com/fabric8io/fabric8-ui

---
This message was auto-generated by https://gitcop.com

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by mindreeper2420** [comment]
@joshuawilson @dlabrecq The latest push includes Mitch's last merged changes for the Space Wizard. Unfortunately, I cannot get past the step of choosing a New Quick Starter Project. In works on Prod, so I'm lost on what the issue is. It just hangs the environment. Can either (or both) of you take a look for me?

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by mindreeper2420** [comment]
@joshuawilson @dlabrecq can either of help out with a console error that is now happening? The Space Wizard will no longer move to selecting the pipeline. This is the error in the snapshot console:
screen shot 2017-06-07 at 12 33 14 pm

**Comment by joshuawilson** [comment]
[test]

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by mitchharpur** [comment]
regarding the pipeline viewer failing  ... not that co.pipeline.steps is undefined ... should be co.buildPipeline.steps ... did something get missed in a merge ?

**Comment by mindreeper2420** [comment]
updated with the fix for the pipeline steps, as well as css fixes with the pipeline layout after conversion to Less.

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by mindreeper2420** [comment]
@joshuawilson I ran this integrated with @SMahil's Planner PR, and all is good. I'm not sure which should be merged first, but everything has been manually checked and verified (existing bugs not withstanding).

**Comment by joshuawilson** [comment]
@mindreeper2420 please get Planner merged and released so that it can be tested with Platform.

File: package.json [hunk 8]
```diff
@@ -164,8 +164,10 @@
     "favicons-webpack-plugin": "0.0.7",
     "file-loader": "0.11.1",
     "gh-pages": "0.12.0",
-    "gulp": "3.9.1",
+    "gulp": "^3.9.1",
```
Comment (joshuawilson) [inline package.json]: to be clear, please remove all of them

**Comment by mindreeper2420** [comment]
@joshuawilson package.json has been updated.

@dlabrecq - I fixed the wrapping of the Codebases row, but was unable to figure out why steps 4 and 5 are duplicated in the Space Wizard. I checked on the spinner, and it is centered in my environment and the latest snapshot build.

**Comment by mindreeper2420** [comment]
@dlabrecq the highlight of both step 4 and step 5 labelled 'build configuration' is currently in master, so I won't be addressing that as part of this PR :-)

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by mindreeper2420** [comment]
closes https://github.com/fabric8io/fabric8-ux/issues/486

**Comment by mindreeper2420** [comment]
closes https://github.com/fabric8io/fabric8-ux/issues/487

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by mindreeper2420** [comment]
fabric8-planner dependency to v0.24.21, in order to consume the flex-to-grid conversion.

**Comment by joshuawilson** [comment]
Your merge failed, you have some bad stuff in the code. If I have time, I'll look at it tonight.

**Comment by joshuawilson** [comment]
oh, that is in the Planner
@SMahil the rebase has some bad stuff in it in the planner code.
cc @michaelkleinhenz

**Comment by SMahil** [comment]
@joshua is this stuff related to merge conflicts??

**Comment by joshuawilson** [comment]
@SMahil yes. When you tried to resolve them you left the artifacts in the code of the old and new versions. You can see it in the details log for ci.centos listed here.

**Comment by SMahil** [comment]
I have raised the PR on that conflicts

**Comment by mindreeper2420** [comment]
I will update `package.json` with a new Planner version once PR https://github.com/fabric8io/fabric8-planner/pull/2010 is merged.

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io

**Comment by joshuawilson** [comment]
@mindreeper2420 it appears that the Planner is not full size.

**Comment by mindreeper2420** [comment]
@joshuawilson - the Planner will not be, as it is clashing with the flex-less view. When the Planner is fixed and run with this PR running integrated, it looks/works as it should.

**Comment by mindreeper2420** [comment]
@dlabrecq can you review with the updates? I believe I addressed all of the requested changes. Thanks!

**Comment by fabric8cd** [comment]
@mindreeper2420 snapshot fabric8-ui is deployed and available for testing at https://fabric8-ui-fabric8-ui-pr-1419-fabric8-ui.int.rdu2c.fabric8.io
